src/app/shared/pipes/format-keyboard-keys.pipe.ts

```diff
@@ -5,6 +5,9 @@ import { Pipe, PipeTransform } from '@angular/core'
 })
 export class FormatKeyboardKeysPipe implements PipeTransform {
   public transform(keys: string[]): string {
+    if (!keys) {
+      return ''
+    }
```

Comment:
Why is this check added?

Reply:
To avoid the undefined exception when it tries to call `.map` on undefined on the line below.

Reply:
A `KeyboardMapping` that for some reason hasn't gotten any Keys.

Reply:
Then that is not the responsibility of this component. The contract for this pipe is that it takes a string array to format. If an `undefined` is passed to it at runtime, then we are using types improperly.

Reply:
Making the `KeyboardInputComponent` able to clear it self has now made the type optional.